[plugins] maxage and all_logs on all plugins

Akrog opened the pull request:
This patch makes 2 options settable globally and per-plugin: `maxage` and `all_logs`.

Examples:

```
sos report --all-logs --maxage=1
sos report -k container_log.all-logs=on,container_log.maxage=1
sos report --all-logs --maxage=1 -k container_log.all-logs=off,container_log.maxage=2
```

This adds flexibility to what users want to retrieve.

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [x] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?
- [x] Are all passwords or private data gathered by this PR [obfuscated](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin#how-to-prevent-collecting-passwords)?

Akrog committed:
[plugins] maxhours and all_logs on all plugins

This patch makes 2 options settable globally and per-plugin: `maxhours`
and `all-logs`.

The `maxhours` option maps to what the code used to call `maxage`, which
has also been renamed to `maxhours` to be consistent.

Examples:

```
sos report --all-logs --maxhours=1
sos report -k container_log.all-logs=on,container_log.maxhours=1
sos report --all-logs --maxhours=1 \
  -k container_log.all-logs=off,container_log.maxhours=2
```

This adds flexibility to what users want to retrieve.

Signed-off-by: Gorka Eguileor <geguileo@redhat.com>

pmoravec commented on sos/report/plugins/__init__.py:
```diff
@@ -1674,6 +1685,9 @@ def add_copy_spec(self, copyspecs, sizelimit=None, maxage=None,
         if self.get_option('since'):
             since = self.get_option('since')
 
+        # User's plugin option argument takes precedence over plugin passed arg
+        maxage = self.get_option('maxage') or maxage
+
```
Shall not we unify the code here? Both manipulations perform the same at the end, the unification would improve clarity of code.

pmoravec commented on sos/report/__init__.py:
```diff
@@ -200,6 +201,12 @@ def add_parser_options(cls, parser):
                                 help="Escapes archived files older than date. "
                                      "This will also affect --all-logs. "
                                      "Format: YYYYMMDD[HHMMSS]")
+        report_grp.add_argument("--maxage", action="store",
+                                dest="maxage", default=None, type=int,
+                                help="Escapes archived files older (according "
+                                "to `mktime`) than this many hours. This will "
+                                "also affect --all-logs (sets the default for "
+                                "all plugins)")
```
Would it be clear for users the option takes values in *hours*? Esp. if `--since` is primarily in days (default expected format is a day, optionally time)? Should not we rename it to, I dont know, `--maxhours`? (that sounds badly to me, I like the `--maxage` more).

I can easily be wrong here. I really dont know what users' perception would be here.

Akrog replied:
I like the suggested change to `maxhours`, as it has the benefit of including the unit in the name and not force people to look at the description to figure it out.

I'll make the change.

pmoravec commented:
I meant that

            since = None
            if self.get_option('since'):
                since = self.get_option('since')
            
            # User's plugin option argument takes precedence over plugin passed arg
            maxhours = self.get_option('maxhours') or maxhours

applies same logical change to either `since` or `maxhours` option. For the sake of code coherence, we should use same construction in either assignment, so let have there either:

            since = self.get_option('since') or None
            
            # User's plugin option argument takes precedence over plugin passed arg
            maxhours = self.get_option('maxhours') or maxhours

or:

            since = None
            if self.get_option('since'):
                since = self.get_option('since')
            
            # User's plugin option argument takes precedence over plugin passed arg
            if self.get_option('maxhours'):
                maxhours = self.get_option('maxhours')

The first construction is more concise, the second more explicit. I like the first more but 2nd is fully OK for sure.

TurboTurtle commented on sos/report/__init__.py:
```diff
@@ -904,6 +911,12 @@ def _set_all_options(self):
                         opt.value = True
 
     def _set_tunables(self):
+        # Set plugin option's defaults to the global argument values
+        for pluginname, plugin in self.loaded_plugins:
+            for optname in plugin.options:
+                if hasattr(self.opts, optname):
+                    plugin.options[optname].value = getattr(self.opts, optname)
```
I'm not seeing why this is needed? We have the defaults for the plugin options set when they're instantiated as a `PluginOpt`, and these shouldn't differ at this point.

TurboTurtle replied:
Nevermind, I see it now. The other options to date have had specific handling to them for default vs. plugin-specific values due to the need to do some dancing with magic numbers. We don't need to do that here, so we're good.

TurboTurtle commented:
In a quick test, this change works for `maxhours`, but does not work for `all_logs` - the plugin only gets the global value, which is due to this in `Plugin`:

https://github.com/sosreport/sos/blob/7b0d5a4355e380e9d662754efe0eb9b786e9f217/sos/report/plugins/__init__.py#L1544-L1564

Akrog commented on sos/report/plugins/__init__.py:
```diff
@@ -585,7 +585,18 @@ def __init__(self, commons):
+                desc='Escapes archived files older (according to `mktime`) '
+                'than this many hours'
+            ),
+            'all_logs': PluginOpt(
```
I'm new to this, but I assumed looking at the [kernel plugin `with-timer` definition](https://github.com/sosreport/sos/blob/e48a41dd6da6178cc04ce6a6d7685814195e4384/sos/report/plugins/kernel.py#L39) and how it is [accessed as `with-timer` and not `with_timer`](https://github.com/sosreport/sos/blob/e48a41dd6da6178cc04ce6a6d7685814195e4384/sos/report/plugins/kernel.py#L144) that there is no translation on the `PluginOpt`.

I looked a bit into the `get_option` and `PluginOpt` and didn't find any translation mechanism so I went with the solution that required fewer code changes.

But it is true that it would be weird to call it `--all-logs` only to have to use `all_logs` in the plugins.

I'll see if I can use `all-logs` everywhere.

Akrog replied:
Done

Akrog committed:
[PluginOpt] Support different parameter name

Current `PluginOpt` code doesn't have the possibility of having a
different name for the parameter and the reference used in the code.

This would be the equivalent in arg parser to not having a `dest`
argument available to us.

This patch proposes a change to allow this using a new `PluginOpt` field
called `param_name`.

This new field will only be used when interfacing with users. In other
words, when showing the options to users (eg: `list_plugins`), when
parsing the command line arguments (eg: `_set_tunables`), and when
reading the config file (this doesn't require code changes because it
leverages the normal `_set_tunables` method).

Signed-off-by: Gorka Eguileor <geguileo@redhat.com>

Akrog commented:
> In a quick test, this change works for `maxhours`, but does not work for `all_logs` - the plugin only gets the global value, which is due to this in `Plugin`:
> 
> https://github.com/sosreport/sos/blob/7b0d5a4355e380e9d662754efe0eb9b786e9f217/sos/report/plugins/__init__.py#L1544-L1564

Done, I implemented it in what I consider the safest way, and it also adds a missing feature: different variable and user facing parameter name for plugin options.

Alternatively the `all-logs` command line argument could be changed to save the parameter in `all-logs` instead of `all_logs` and all code changed to access it with `getattr` instead of using `.all_logs`.

TurboTurtle commented:
Sorry for the silence here, cycling around on it now.

The `all-logs` vs `all_logs` convention is due to python variable identifiers not allowing hyphen characters. We aligned to using hyphens in the user-facing CLI because that's typically what users expect, and it is the prevailing syntax for argparse and what it does automatically for name conversions.

Internally, developers just need to know that hyphens become underscores when accessing the associated variables, which is the case for any project using argparse and directly accessing the parsed options.

So, we're good with the original intent of the PR - expanding all-logs and maxage/maxhours to all plugins on a potentially individual basis, but we don't need to mess around with internal references to option names.

pmoravec commented:
I am not sure of the sttaus of this PR - are we waiting for a change from author, or for a review? (I plan to review the current PR within a day or two, though it would be great if PR is rebased to current main - there are conflicts..)

pmoravec commented:
I like it. Maybe worth adding a (report_)test? (I can come up with something if preferred)

arif-ali commented:
@Akrog any update on this PR at all?
